Option to enable computed functions in session endpoints

sravfeyn opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

Adds a new option in form settings when session endpoints toggle is ON.

## Technical Summary
https://dimagi-dev.atlassian.net/browse/SC-1838

## Feature Flag
SESSION_ENDPOINTS

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.
-->
This introduces an option in settings, and when the option is used, the suite is generated differently per the mobile suite spec. Tests are added to make sure the suite is correct. So there should be no issues

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Tests added

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->

Planning to put it on staging with basic developer testing

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->


<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

sravfeyn committed:
Option to enable computed functions in session endpoints

sravfeyn commented on corehq/apps/app_manager/tests/test_suite_session_endpoints.py:
```diff
@@ -315,3 +316,135 @@ def test_shadow_module(self):
+        builder.new_question(name='name', label='Name')
+        self.form.source = builder.tostring(pretty_print=True).decode('utf-8')
+
+    def test_without_computed(self):
```
@shubham1g5 Hey Shubham, I have added two tests [with](https://github.com/dimagi/commcare-hq/pull/30707/files#diff-5e51d180dd0fde10d5decd2e10a001042c880b1bdace14e217a7e821a5e11f5bR393) and without the computed datums to test the resulting suite xml. Can you verify that the endpoint xml is as you specified.

shubham1g5 replied:
looks good to me, though is it possible to deploy this on staging for some testing.

sravfeyn replied:
This should be on staging now. Let me know when you get a chance to test.

shubham1g5 replied:
@orangejenny There seems to be an issue how this PR is interacting with the `remote-request` work you did earlier. I am testing this on stagaing and in the [suite file](https://staging.commcarehq.org/a/shubhamgoyaltest/apps/download/9261779eff8846d485e421dda7861b84/suite.xml), the `sms_form` endpoint is adding up a `command` in stack `claim_command.sms_form.datum_test` definition for which is not present in the suite file. Probably we should not treat the computed datums as case ids here ?

orangejenny replied:
Yeah, that sounds likely. @sravfeyn you'll want to ignore computed datums in two places: [here](https://github.com/dimagi/commcare-hq/blob/178f5c43df87e329595b13e0d867e807b0f015d9/corehq/apps/app_manager/suite_xml/post_process/endpoints.py#L46-L49) is where the claim frames are added to the endpoint stack, and [here](https://github.com/dimagi/commcare-hq/blob/178f5c43df87e329595b13e0d867e807b0f015d9/corehq/apps/app_manager/suite_xml/post_process/remote_requests.py#L336-L339) is where the remote-requests for those claim are created.

Is this going to be smoke tested on formplayer / web apps, too?

shubham1g5 replied:
> Is this going to be smoke tested on formplayer / web apps, too?

Do you mean testing for any regressions because of this ? As far as this computed datum feature is concerned, I am only planning to test this out on mobile which should be able to confirm these changes are correct and results into right config.

sravfeyn replied:
https://github.com/dimagi/commcare-hq/pull/30707/commits/26f419ea9e175a25e8d6c21d2d3b0e1d44ec0236

orangejenny replied:
just regressions

shubham1g5 replied:
Offlined with @orangejenny and decided that it should be fine to not do any regression testing since no one is actively using any of session endpoint stuff currently both on mobile and Formplayer.

sravfeyn committed:
Add html file

proteusvacuum commented on corehq/apps/app_manager/suite_xml/post_process/endpoints.py:
```diff
@@ -63,10 +63,20 @@ def _make_session_endpoint(self, module, form=None):
+            if child.requires_selection:
+                return True
+            if form:
+                return child.id in form.function_datum_endpoints or []
```
This is going to return `[]` if the id isn't in `function_datum_endpoints`. I think you want:
```suggestion
                return child.id in (form.function_datum_endpoints or [])
```
if you are trying to protect against `function_datum_endpoints` being Falsey

esoergel replied:
Won't jsonobject already default a `StringListProperty` to an empty list if the stored object doesn't have it defined?

sravfeyn replied:
Thanks Farid. Fixed.

> Won't jsonobject already default a StringListProperty to an empty list if the stored object doesn't have it defined?

True. It gets populated via user input, so there could be a case where this might get set to None, which this will handle.

sravfeyn committed:
Correction

sravfeyn committed:
exclude computed datums in claim

sravfeyn commented:
Hey @shubham1g5 I am dropping offline for PTO. Feel free to merge this whenever, since this PR is approved and I have addressed your feedback. If there is anything else please let Mary/Glitter know.
